src/Handlers/ShortcodeParametersHandler.php

```diff
@@ -0,0 +1,159 @@
+     * @var TransformerInterface
+     */
+    private $serviceTransformer;
+
+    /**
+     * MainComponentHandler constructor.
+     *
+     * @since [*next-version*]
+     *
+     * @param int                        $cartPageId                 Cart page ID.
+     * @param SelectCapableInterface     $serviceSelectResourceModel Resource model for selecting services.
+     * @param ExpressionBuilderInterface $expressionBuilder          Expression builder.
+     * @param TransformerInterface       $serviceTransformer         Service transformer.
+     */
+    public function __construct($cartPageId, $serviceSelectResourceModel, $expressionBuilder, $serviceTransformer)
+    {
+        $this->cartPageId                 = $cartPageId;
+        $this->serviceSelectResourceModel = $serviceSelectResourceModel;
+        $this->expressionBuilder          = $expressionBuilder;
+        $this->serviceTransformer         = $serviceTransformer;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function __invoke()
+    {
+        /* @var $event EventInterface */
+        $event = func_get_arg(0);
+
+        if (!($event instanceof EventInterface)) {
+            throw $this->_createInvalidArgumentException(
+                $this->__('Argument is not an event instance'), null, null, $event
+            );
+        }
+
+        $event->setParams($this->_handleParameters($event->getParams()));
+    }
+
+    /**
+     * Render booking holder.
+     *
+     * @since [*next-version*]
+     *
+     * @param $params
+     *
+     * @return array
+     */
+    protected function _handleParameters($params = [])
+    {
+        $params['redirectUrl'] = $this->_getRedirectUrl();
+
+        if (isset($params['service'])) {
+            $service = $this->_getService((int) $params['service']);
+            unset($params['service']);
+            if ($service) {
+                $params['service'] = $service;
+            }
+        }
+
+        return $params;
+    }
+
+    /**
+     * Get cart URL on which customer will be redirected after successfull booking creation.
+     *
+     * @since [*next-version*]
+     *
+     * @return string Cart URL to redirect user on.
+     */
+    protected function _getRedirectUrl()
+    {
+        return get_permalink($this->cartPageId);
+    }
+
+    /**
+     * Get service by service ID.
+     *
+     * @since [*next-version*]
+     *
+     * @param int $serviceId Service ID.
+     *
+     * @return stdClass|null Service data if service is found.
+     */
+    protected function _getService($serviceId)
+    {
+        $b = $this->expressionBuilder;
+
+        $condition = $b->and(
+            $b->eq(
+                $b->ef('service', 'id'),
+                $b->lit($serviceId)
+            )
+        );
+        $services = $this->serviceSelectResourceModel->select($condition);
+
+        return isset($services[0]) ? $this->serviceTransformer->transform($services[0]) : null;
```

Comment:
The job of this method is to retrieve a service. The job of the transformer is to make the service data look good for the API. These are two separate concerns, and therefore the transformation shouldn't happen here.